app/controllers/autocomplete_controller.rb

```diff
@@ -91,16 +91,30 @@ def associated_tags
   def noncanonical_tag
     search_param = params[:term]
     raise "Redshirt: Attempted to constantize invalid class initialize noncanonical_tag #{params[:type].classify}" unless Tag::TYPES.include?(params[:type].classify)
+
     tag_class = params[:type].classify.constantize
-    render_output(tag_class.by_popularity
-                      .where(["canonical = 0 AND name LIKE ?",
-                              '%' + search_param + '%']).limit(10).map(&:name))
+    one_tag = tag_class.where(canonical: false, name: search_param)
+    # Is there a tag which is just right ( this is really for testing )
+    match = []
+    match = [one_tag.first.name] if one_tag.count == 1
+
+    word_list = search_param.split
+    last_word = word_list.pop
+    search_list = word_list.map { |w| { term: { name: { value: w, case_insensitive: true } } } } + [{ prefix: { name: { value: last_word, case_insensitive: true } } }]
```

Comment:
Out of curiosity, what is this request inspired by? I couldn't find anything similar in the current code or the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-suggesters.html).

Reply:
From staging:

```
3.0.5 :133 > example("kaeinui")
 => ["Kaeinui Chwihyang | Personal Preference"]
 =>
["The Amazing Adventures of Kavalier & Clay",
 "Kare Kano",
 "Kaiji (anime)",
 "Amazing Adventures of Kavalier & Clay - Chabon",
 "Kabuki",
 "Kal Ho Naa Ho",
 "Kate Shugak Series",
 "Fionavar Tapestry - Guy Gavriel Kay",
 "Amazing Adventures of Kavalier & Clay - Michael Chabon",
 "Karakuridôji Ultimo/機巧童子ULTIMO"]
3.0.5 :135 > example("kaein")
 => ["Kaeinui Chwihyang | Personal Preference"]
3.0.5 :136 > example("kaei")
 => ["Kaeinui Chwihyang | Personal Preference", "Aoi Shiro - Kaeishou"]
3.0.5 :137 > example("kae")
 =>
["Kaeinui Chwihyang | Personal Preference",
 "Kael'thas Sunstrider/Illidan Stormrage",
 "Kael'thas/Illidan",
 "Kaelthas - Fandom",
 "Albox - Eien o Kaetara",
 "Kael'thas Sunstrider (character)",
 "Kael'thas Sunstrider - Fandom",
 "Aoi Shiro - Kaeishou",
 "Illidan x Kael'thas",
 "Illidan/Kael'thas - relationship"]
3.0.5 :138 > example("Kaeinui Chwihyang | Personal Preference")
 => ["Kaeinui Chwihyang | Personal Preference"]
```

Am I holding suggestions badly ?

```
curl -X POST 'localhost:9200/ao3_staging_tags/_search?pretty' -H 'Content-Type: application/json' '-d
{
  "suggest": {
    "my-suggest-1" : {
      "text" : "Kae",
      "term" : {
        "field" : "name"
      }
    }
  }
}
'
{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 0,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "suggest" : {
    "my-suggest-1" : [
      {
        "text" : "kae",
        "offset" : 0,
        "length" : 3,
        "options" : [ ]
      }
    ]
  }
}
```

Reply:
Okay, I've tried using the [Completion Suggester](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-suggesters.html#completion-suggester): https://github.com/ceithir/otwarchive/commit/9cd6bbb228504278678d8a54e47da9768a6181dd

... Raised an eyebrow at the fact it [only supports](https://stackoverflow.com/questions/48304499/elasticsearch-completion-suggester-not-working-with-whitespace-analyzer#answer-54080114) completion from the left (there's possibly a way to work around that with regex, but haven't delved that deep)...

And had said eyebrow curling at what needed to be done to add the equivalent of a filter query.

It might indeed not be what's best for our current usage in spite of it theoretically be ES auto-complete feature.